internal/metastore/db/dao/collection.go

```diff
@@ -0,0 +1,124 @@
+package dao
+
+import (
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/metastore/db/dbmodel"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+	"go.uber.org/zap"
+	"gorm.io/gorm"
+)
+
+type collectionDb struct {
+	db *gorm.DB
+}
+
+func (s *collectionDb) GetCidTs(tenantID string, collectionID typeutil.UniqueID, ts typeutil.Timestamp) (*dbmodel.Collection, error) {
+	if ts > 0 {
+		tx := s.db.Model(&dbmodel.Collection{}).Select("collection_id, ts")
+
+		if tenantID != "" {
+			tx = tx.Where("tenant_id = ?", tenantID)
+		}
+
+		var col dbmodel.Collection
+
+		err := tx.Where("collection_id = ? AND ts > 0 and ts <= ?", collectionID, ts).Order("ts desc").Limit(1).Find(&col).Error
+		if err != nil {
+			log.Error("get collection ts failed", zap.Int64("collID", collectionID), zap.Uint64("ts", ts), zap.Error(err))
+			return nil, err
+		}
+
+		return &col, nil
+	}
+
+	return &dbmodel.Collection{CollectionID: collectionID, Ts: 0}, nil
+}
+
+func (s *collectionDb) ListCidTs(tenantID string, ts typeutil.Timestamp) ([]*dbmodel.Collection, error) {
+	tx := s.db.Model(&dbmodel.Collection{}).Select("collection_id, ts")
+
+	if tenantID != "" {
+		tx = tx.Where("tenant_id = ?", tenantID)
+	}
+
+	var err error
+	var r []*dbmodel.Collection
+
+	if ts > 0 {
+		err = tx.Select("collection_id, MAX(ts) ts").Where("ts > 0 and ts <= ?", ts).Group("collection_id").Find(&r).Error
+	} else {
+		err = tx.Select("collection_id, MAX(ts) ts").Where("ts = 0").Group("collection_id").Find(&r).Error
```

Comment:
If `ts = 0`, it seems that no record will be hit, since anytime we will ts to create/drop a collection.